### ISSUE TITLE
Action Buttons not appearing on iOS until notification made explicitly rich by adding picture

### ISSUE THREAD

**Issue by XenorPLxx** [OneSignal/OneSignal-iOS-SDK#430]
**Description:**

Hello!

I've setup OneSignal push notifications in my React Native app that is build for both Android and iOS.

I'm trying to send basic notification from OneSignal dashboard that includes buttons. While my app is in foreground, everything looks fine - I see buttons in payload, so I display them in in-app notification.

When application is in background or closed, buttons are handled as a part of system notification. On Android it works fine, but on iOS I don't see the buttons (even tho they are in payload in my `opened` `eventListener`.

I rechecked if native setup done correctly on iOS and I've started to play with different settings in the OneSignal dashboard. After I've added picture to iOS notification (with buttons setup), I've got BOTH picture and buttons. And after that, even without picture, I've started getting buttons correctly on iOS.

It all worked until I've tried removing app from my phone. After reinstallation buttons disappeared and again, sending picture once made them available in later notifications.

React Native library is just a SDK wrapper for handling notifications, hence I think this is the best place to put that issue in.

**Environment**
1. What version of the iOS SDK are you using?
react-native 3.2.7 includes iOS SDK 2.8.8 (per https://github.com/geektimecoil/react-native-onesignal/releases)
2. How did you add the SDK to your project (eg. cocoapods)
Linking
3. iOS version
12.0, 12.1

**Steps to Reproduce Issue:**

1. Setup notifications
2. Try sending buttons with notification
3. Observe, if buttons are available in system notification (app in background)
4. If not, try adding picture to make notification Rich
5. Again observe, if buttons are available in system notification (app in background)

**Comment by Nightsd01** [comment]
@XenorPLxx Can you share how you set up your Notification Service Extension? 

Also, in iOS, buttons **will not show up at all* unless you force-press on the notification. This is **not** a bug. It is iOS system behavior. You need to force-press/pull down to see the buttons. Just want to make sure that's not the issue.

**Comment by XenorPLxx** [comment]
Hiya.

Thanks for the response. I'm aware that I have to use 3D touch to make buttons appear on iOS, that is not the case. Buttons do not appear until I send at least one notification with picture included.

For setup I've followed https://documentation.onesignal.com/v5.0/docs/react-native-sdk-setup, including adding Notification Service Extension with Objective-C code.

For sending notifications I've tried both OneSignal dashboard (although buttons are part of Android notification there) and API by following documentation, which states that buttons work for both Android and iOS.

I can upload screens of the steps mentioned in setup documentation tomorrow when I'll have access to my workstation if needed.

**Comment by Nightsd01** [comment]
Can you send your `NotificationService.m` and `NotificationService.h` files?

Can you also post a screenshot of the Xcode Project Settings for your OneSignalNotificationServiceExtension target, especially the `Link Binary with Libraries` section, and can you also post your Podfile?

**Comment by Nightsd01** [comment]
Can you reproduce this behavior in our example project location in `/examples/RNOneSignal` folder from the react-native repo?

If so, can you please zip up the project and email a dropbox/etc. link to brad@onesignal.com so I can try to reproduce this issue? You don't have to send us your full app - just the demo app after you install dependencies and reproduce the issue.

It would also be helpful for you to give me a Notification ID for one of the notifications that exhibited the problem - even if it's an exact duplicate of a notification that works later on.

It's a bizarre issue - we've never heard of anything like this being reported before - so I'd definitely like to see exactly how you've got things configured.

**Comment by XenorPLxx** [comment]
Sure, I'll try to prepare reproduction app over the weekend and I'll get back to you.

**Comment by XenorPLxx** [comment]
Hello! 

I'm happy to inform that I have the repro ready and issue still appears (well, maybe not 'happy' happy, but you know what I mean :D).

After downloading example app I:
1. Installed dependencies using `yarn`
2. Ran third-party glog configurations and relinked libfishhook.a (fix for `react-native 0.55.4` Xcode 10 bug)
3. Changed bundle id and setup automatic signing to use my dev account
4. Setup new push cert (following https://documentation.onesignal.com/docs/generate-an-ios-push-certificate)
5. Setup Notification Service Extension (following https://documentation.onesignal.com/v5.0/docs/react-native-sdk-setup)
4. Setup new app in OneSignal (including iOS cert setup) and changed id in `OneSignal.init` to my new OneSignal app id
5. Setup message template that includes button
6. Built and installed app through Xcode 10 Beta 2 on iPhone X running iOS 12.1 Beta 2 (my colleague tried also stable iOS 12 on iPhone X with the original app)
7. Registered email in example app and checked, if user is visible in OneSignal dashboard
8. Sent notificaton with buttons while app in background - no buttons (screen below)
![img_0474](https://user-images.githubusercontent.com/6712823/46573402-d45ce280-c994-11e8-9e52-c33dab7d0154.PNG)
9. Sent notification with buttons AND iOS picture - buttons appeared (screen below)
![img_0475](https://user-images.githubusercontent.com/6712823/46573405-d757d300-c994-11e8-853f-1f73834a36fd.PNG)
10. Sent notificaton with buttons while app in background - buttons appeared (screen below)
![img_0476](https://user-images.githubusercontent.com/6712823/46573408-db83f080-c994-11e8-87da-f1b33b0737c0.PNG)
11. Removed and installed application again
12. Sent notificaton with buttons while app in background - no button (screen below)
![img_0477](https://user-images.githubusercontent.com/6712823/46573410-dfb00e00-c994-11e8-952f-52c272be5888.PNG)


Ziped repro app: https://synology.dawidkarczewski.pl/sharing/AerCCMgV8 (URL will expire at 31.12.2018)

**Comment by Nightsd01** [comment]
Hey @XenorPLxx thanks for all the details! 

I looked into the issue and it looks like your project has the Deployment Target for the OneSignalNotificationServiceExtension set to 12.1.

You will definitely want to change this to 10.0 and that fixes the issue.

I will need to do some further investigation to figure out why building the extension for 12.0+ appears to crash the extension before it can attach buttons/etc. But it works perfectly well with 10.0.

![fix](https://user-images.githubusercontent.com/8186991/46589001-b6a58100-ca59-11e8-8961-698cc62dac0a.jpg)

**Comment by XenorPLxx** [comment]
Hiya!

Thanks for the response.
Sadly, it didn't do the trick for me. I've tried both repro app and original app we had problems with and both don't display buttons after removing and rebuilding with Deployment Target = 10.0 for Notification Extension.

Here's example notification that didn't show buttons:
![img_68d0138b2b0b-1](https://user-images.githubusercontent.com/6712823/46600610-f7010f80-caea-11e8-9efb-e4dfae253ba3.jpeg)

**Comment by Nightsd01** [comment]
Hi @XenorPLxx I was able to reproduce the issue...until I lowered the deployment target. Now I cannot reproduce, even uninstalling + reinstalling the example app you sent. 

I am thinking this may be an issue in the Xcode + iOS betas as these distributions often have bugs. Can you try downgrading to the public releases of Xcode 10 & iOS 12.0 and lower the deployment target of the extension service and tell me if the issue still occurs?

**Comment by Nightsd01** [comment]
Also, here is an easier way to debug this. Instead of attaching buttons, try replacing the contents of your `NotificationService.m` with this instead:

```objc
#import <RCTOneSignalExtensionService.h>

#import "NotificationService.h"

@interface NotificationService ()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNNotificationRequest *receivedRequest;
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation NotificationService

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
  self.receivedRequest = request;
  self.contentHandler = contentHandler;
  self.bestAttemptContent = [request.content mutableCopy];
  
  self.bestAttemptContent.body = [@"[Modified] " stringByAppendingString:self.bestAttemptContent.body];
  
  self.contentHandler(self.bestAttemptContent);
}

- (void)serviceExtensionTimeWillExpire {
  // Called just before the extension will be terminated by the system.
  // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
  
  [RCTOneSignalExtensionService serviceExtensionTimeWillExpireRequest:self.receivedRequest withMutableNotificationContent:self.bestAttemptContent];
  
  self.contentHandler(self.bestAttemptContent);
}

@end

```

Now, when you do that, whenever you receive a rich push notification (with buttons/photos/etc.) the body of the push notification will have `[Modified]` added to the front of the message. Keep in mind that iOS only knows a push notification is a rich notification if `mutable-content` is set to `true`

**Comment by XenorPLxx** [comment]
I'm not really keen on resetting my iPhone to downgrade iOS, so I'll try some device when I'm at work tomorrow that has stable iOS 12 or lower.

I've tried second solution you've provided (with target lowered to 10.0) and as you can see the `[Modified]` text is there, but buttons are not:

![img_0479](https://user-images.githubusercontent.com/6712823/46627146-d4e2ae00-cb39-11e8-8dd1-b785b8c8c525.PNG)

![img_0480](https://user-images.githubusercontent.com/6712823/46627147-d4e2ae00-cb39-11e8-9ec0-2c17f4a5dbd5.PNG)

**Comment by Nightsd01** [comment]
@XenorPLxx Keep in mind that code I posted for `NotificationService.m` would definitely stop buttons from working even if they had been working in the first place.

I am pretty certain this is a bug in Xcode 10 beta or iOS 12.1 beta. I will attempt to reproduce.

**Comment by XenorPLxx** [comment]
The only test I could do without access to additional iOS devices was to copy over support files for iOS 12.1 to public release XCode 10 and build from there, but i didn't do any change.

I've also tried uncommenting 2 lines in original `NotificationService.m` (25 & 26) per in-file comment suggestion and I've got `[Modified]` text, but buttons didn't appear.

**Comment by Nightsd01** [comment]
Yeah, even with iOS 12 beta + Xcode 10 beta, I cannot reproduce your issue.

Just to make _absolutely_ sure you are setting the deployment target correctly for the **extension** target and not the **app** target can you zip up the demo project as you currently have it and send it to me?

**Comment by XenorPLxx** [comment]
Changing target for app target is blocked in my XCode.

Here's the current app zipped: https://synology.dawidkarczewski.pl/sharing/yHH9UbSQd (URL will expire at 31.12.2018)

**Comment by Nightsd01** [comment]
I've had another idea. Try this. Try replacing the contents of your `NotificationService.m` with this.

Then, send a notification with buttons. We want to know two things: (A) Do the buttons appear? And (B) Does the title of the notification show up as [Modified 1] or [Modified 2], or neither?

```objc
#import <RCTOneSignalExtensionService.h>

#import "NotificationService.h"

@interface NotificationService ()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNNotificationRequest *receivedRequest;
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation NotificationService

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
  self.receivedRequest = request;
  self.contentHandler = contentHandler;
  self.bestAttemptContent = [request.content mutableCopy];
  
  self.bestAttemptContent.body = [@"[Modified 1] " stringByAppendingString:self.bestAttemptContent.body];
  
  [RCTOneSignalExtensionService didReceiveNotificationRequest:self.receivedRequest withContent:self.bestAttemptContent];
  
   self.bestAttemptContent.body = [@"[Modified 2] " stringByAppendingString:self.bestAttemptContent.body];
  
  self.contentHandler(self.bestAttemptContent);
}

- (void)serviceExtensionTimeWillExpire {
  // Called just before the extension will be terminated by the system.
  // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
  
  [RCTOneSignalExtensionService serviceExtensionTimeWillExpireRequest:self.receivedRequest withMutableNotificationContent:self.bestAttemptContent];
  
  self.contentHandler(self.bestAttemptContent);
}

@end
```

**Comment by XenorPLxx** [comment]
Here are the result, both text are there, no buttons:

Fresh install:
![img_0481](https://user-images.githubusercontent.com/6712823/46632411-db2d5600-cb4a-11e8-823b-7f04c79a436d.PNG)
![img_0482](https://user-images.githubusercontent.com/6712823/46632412-dbc5ec80-cb4a-11e8-9d41-a40480736b4c.PNG)

Picture attached:
![img_0483](https://user-images.githubusercontent.com/6712823/46632410-db2d5600-cb4a-11e8-873b-53add7a318b3.PNG)
![img_0484](https://user-images.githubusercontent.com/6712823/46632409-da94bf80-cb4a-11e8-9e91-f083353503b2.PNG)

**Comment by XenorPLxx** [comment]
I've just remembered that earlier today I've changed extension target in the original app that has the same problem and tested dev release version - it also doesn't show buttons.

Thanks to that I think we can exclude Xcode beta problems - to build that version we use Bitrise and their Xcode 9.4.1 stack.

**Comment by Nightsd01** [comment]
I have done some more testing and I have been able to reliably reproduce this issue in iOS 12, both the public release and the beta.

I've done some debugging and I've also been able to verify that our SDK is doing _exactly_ what it should be doing. It's correctly registering a `UNNotificationCategory`, registering it with the current `UNUserNotificationCenter` and registering your buttons for that category, and it is also setting the `categoryIdentifier` on the `UNNotificationContent` 

This is everything the system needs to add the button to the notification, but it still doesn't get added.

I will be reporting this as a bug with iOS, and will be adding the radar issue # here.

**Comment by Nightsd01** [comment]
In fact this isn't just occurring with React-Native, it's occurring with our normal iOS SDK release as well. Thanks for reporting this and good catch!

**Comment by XenorPLxx** [comment]
Always glad to help ^^
Thanks for your time so far. I hope that resolving this issue over on Apple side will be supported so well as it is here in OneSignal :).

**Comment by stevesweets** [comment]
Have experienced this issue as well with a react-native build; Finding this report was helpful as i've been able to resolve it with adding an attachment.

**Comment by neo125874** [comment]
@Nightsd01 hi, any further update or workaround could resolve this issue?  thanks.

**Comment by Nightsd01** [comment]
@neo125874 apparently this issue has been reported before I created my bug report, Apple told me they’re aware of the issue and are working on it. I don’t have an ETA at the moment. 

It is quite an aggravating bug. I was able to reproduce even without our SDK. I will update this issue when I hear anything further.

**Comment by User2004** [comment]
https://webappcodes.com/tag/notification

Here is a link for integrate push notification in ios

**Comment by mcontin** [comment]
Any info if this bug has been solved or if it still affects newer releases of iOS 12?

**Comment by rgomezp** [comment]
I haven't heard any recent complaints about this issue. I think it is safe to close. I will leave it open for another week to see if anyone reports anything !mcontin

**Comment by XenorPLxx** [comment]
@GaborWnuk you might want to recheck if issue is still valid :arrow_up:

**Comment by rgomezp** [comment]
Closing due to no response

**Comment by 7freaks-otte** [comment]
Anyone still experiencing this except me?

I updated everything to the newest versions (Xcode 11, iOS 13.1.2), have reimplemented OneSignal with Groups and ServiceExtension (target 10.0) several times, even recreated my app from scratch, double/triple/milliontimes checked every step in the documentation, I just don't get the action buttons to work on iOS.

Action buttons work perfectly on Android, they even work on iOS if I attach a picture, but as soon as I remove the picture again, the buttons are gone as well..

**Comment by Bathant** [comment]
Is there any updates for this issue? Is it fixed?

**Comment by rgomezp** [comment]
@Bathant ,
Howdy.

I believe this was fixed by #597 

You may be experiencing the issue reported in #740

**Comment by scrummitch** [comment]
If anyone finds this, i still have this issue in another product in 2022, iOS 16. 

I firstly solved this by adding a sleep(1) function after i added my category, then listing categories again. Which makes me think there is some disk or synchronisation step in the background happening in the OS. 

See below for my better solution:

```swift
 func setCategories(newCategory: UNNotificationCategory) async -> Bool {
        return await withCheckedContinuation {continuation in
            UNUserNotificationCenter.current().getNotificationCategories { categories in
                var allCategories: Set<UNNotificationCategory> = categories
                allCategories.insert(newCategory)
            
                UNUserNotificationCenter.current().setNotificationCategories(allCategories)
                
                UNUserNotificationCenter.current().getNotificationCategories { categories in
                    continuation.resume(returning: true)
                }
            }
        }
    }
```
